### PR TITLE
[Behats] Rework address book tests

### PR DESCRIPTION
**Pull request by Tomanhez**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

### CODE REVIEW

File: src/Sylius/Behat/Page/Shop/Account/AddressBook/CreatePage.php [hunk 1]
```diff
@@ -42,53 +36,38 @@ public function fillAddressData(AddressInterface $address)
...
         return array_merge(parent::getDefinedElements(), [
-            'add_button' => 'button:contains("Add")',
-            'city' => '#sylius_address_city',
-            'country' => '#sylius_address_countryCode',
-            'first_name' => '#sylius_address_firstName',
-            'last_name' => '#sylius_address_lastName',
-            'postcode' => '#sylius_address_postcode',
-            'street' => '#sylius_address_street',
+            'add_button' => '[data-test-add-address]',
+            'city' => '[data-test--city]',
```
Comment (AdamKasp) [inline src/Sylius/Behat/Page/Shop/Account/AddressBook/CreatePage.php]:
```suggestion
            'city' => '[data-test-city]',
```
and below

File: src/Sylius/Behat/Page/Shop/Account/AddressBook/IndexPage.php [hunk 2]
```diff
@@ -41,97 +35,72 @@ public function getAddressesCount()
...
-    /**
-     * {@inheritdoc}
-     */
-    public function editAddress($fullName)
+    public function editAddress(string $fullName): void
     {
-        $addressToEdit = $this->getAddressOf($fullName);
-        $addressToEdit->findLink('Edit')->press();
+        $addressToEdit = $this->getAddressPageOf($fullName);
+        $addressToEdit->find('css', '[data-test-edit-button] [data-test-button]')->press();
```
Comment (AdamKasp) [inline src/Sylius/Behat/Page/Shop/Account/AddressBook/IndexPage.php]: IMO you should add information what the second button is that, and do we need both attributes?

File: src/Sylius/Behat/Page/Shop/Account/AddressBook/IndexPage.php [hunk 3]
```diff
@@ -41,97 +35,72 @@ public function getAddressesCount()
...
     }
 
-    /**
-     * {@inheritdoc}
-     */
-    public function getFullNameOfDefaultAddress()
+    public function getFullNameOfDefaultAddress(): string
     {
-        $fullNameElement = $this->getElement('default_address')->find('css', 'address > strong');
+        $fullNameElement = $this->getElement('default_address')->find('css', '[data-test-name]');
```
Comment (AdamKasp) [inline src/Sylius/Behat/Page/Shop/Account/AddressBook/IndexPage.php]:
```suggestion
        $fullNameElement = $this->getElement('default_address')->find('css', '[data-test-full-name]'); 
```
?

File: src/Sylius/Behat/Page/Shop/Account/AddressBook/IndexPage.php [hunk 4]
```diff
@@ -41,97 +35,72 @@ public function getAddressesCount()
...
-    /**
-     * @param string $fullName
-     *
-     * @return NodeElement|null
-     */
-    private function getAddressOf($fullName)
+    private function getAddressOf(string $fullName): ?NodeElement
     {
-        return $this->getElement('addresses')->find('css', sprintf('div.address:contains("%s")', $fullName));
+        return $this->getElement('addresses')->find('css', sprintf('[data-test-address]:contains("%s")', $fullName));
```
Comment (AdamKasp) [inline src/Sylius/Behat/Page/Shop/Account/AddressBook/IndexPage.php]:
```suggestion
        return $this->getElement('addresses')->find('css', sprintf('[data-test-address="%s"]', $fullName));
```
and add in html : `sylius_test_html_attribute('address', <variable with full name :D>)`

Comment (AdamKasp) [reply src/Sylius/Behat/Page/Shop/Account/AddressBook/IndexPage.php]: ofc here and everywhere where we have a similar case

File: src/Sylius/Behat/Page/Shop/Account/AddressBook/UpdatePage.php [hunk 5]
```diff
@@ -90,37 +69,30 @@ public function selectCountry($name)
...
         return array_merge(parent::getDefinedElements(), [
-            'city' => '#sylius_address_city',
-            'country' => '#sylius_address_countryCode',
-            'first_name' => '#sylius_address_firstName',
-            'last_name' => '#sylius_address_lastName',
-            'postcode' => '#sylius_address_postcode',
-            'province_name' => 'input[name="sylius_address[provinceName]"]',
-            'province_code' => 'select[name="sylius_address[provinceCode]"]',
-            'save_button' => 'button:contains("Save changes")',
+            'city' => '[data-test--city]',
```
Comment (AdamKasp) [inline src/Sylius/Behat/Page/Shop/Account/AddressBook/UpdatePage.php]:
```suggestion
            'city' => '[data-test-city]',
```
and below

File: src/Sylius/Behat/Page/Shop/Account/AddressBook/IndexPage.php (hunk 2 above)
Comment (GSadee) [inline src/Sylius/Behat/Page/Shop/Account/AddressBook/IndexPage.php]: Couldn't you use `getElement` method here?

File: src/Sylius/Behat/Page/Shop/Account/AddressBook/IndexPage.php (hunk 3 above)
Comment (GSadee) [reply src/Sylius/Behat/Page/Shop/Account/AddressBook/IndexPage.php]: Use `getElement` method

File: src/Sylius/Behat/Page/Shop/Account/AddressBook/IndexPage.php [hunk 6]
```diff
@@ -41,97 +35,72 @@ public function getAddressesCount()
...
+    }
+
+    private function getAddressPageOf(string $fullName): ?NodeElement
+    {
+        return $this->getElement('addresses')->find('css', sprintf('[data-test-address-page]:contains("%s")', $fullName));
```
Comment (GSadee) [inline src/Sylius/Behat/Page/Shop/Account/AddressBook/IndexPage.php]: Use only one element here without an unnecessary second `find`

File: src/Sylius/Bundle/ShopBundle/Resources/views/Account/AddressBook/_defaultAddress.html.twig [hunk 7]
```diff
@@ -1,14 +1,14 @@
 {% import '@SyliusUi/Macro/buttons.html.twig' as buttons %}
 
 <div class="ui top attached styled header">{{ 'sylius.ui.your_default_address'|trans }}</div>
-<div class="ui attached segment" id="sylius-default-address">
-    <div class="ui stackable two column grid">
+<div class="ui attached segment" id="sylius-default-address" {{ sylius_test_html_attribute('default-address') }}>
+    <div class="ui stackable two column grid" {{ sylius_test_html_attribute('address ', address.firstName ~ ' ' ~ address.lastName) }}>
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Account/AddressBook/_defaultAddress.html.twig]: and also
```suggestion
    <div class="ui stackable two column grid" {{ sylius_test_html_attribute('address', address.firstName ~ ' ' ~ address.lastName) }}>
```

File: src/Sylius/Bundle/ShopBundle/Resources/views/Account/AddressBook/_item.html.twig [hunk 8]
```diff
@@ -2,13 +2,13 @@
 
 <div class="eight wide column">
     <div class="ui segment address">
-        <div class="ui stackable two column grid">
+        <div class="ui stackable two column grid" {{ sylius_test_html_attribute('address ', address.firstName ~ ' ' ~ address.lastName) }}>
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Account/AddressBook/_item.html.twig]: same like in previously comment

File: src/Sylius/Bundle/ShopBundle/Resources/views/Account/AddressBook/_defaultAddress.html.twig (hunk 7 above)
Comment (AdamKasp) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Account/AddressBook/_defaultAddress.html.twig]: what do you think about:
```suggestion
    <div class="ui stackable two column grid" {{ sylius_test_html_attribute('address ', "%s %s"|format(address.firstName, address.lastName) }}>
```

File: src/Sylius/Bundle/ShopBundle/Resources/views/Common/_address.html.twig [hunk 9]
```diff
@@ -1,10 +1,10 @@
 {% import "@SyliusUi/Macro/flags.html.twig" as flags %}
 
-<address>
+<address {{ sylius_test_html_attribute('address-context', address.firstName ~ ' ' ~ address.lastName) }}>
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Common/_address.html.twig]: and here and bottom :)

File: src/Sylius/Bundle/ShopBundle/Resources/views/Account/AddressBook/_defaultAddress.html.twig [hunk 10]
```diff
@@ -1,14 +1,14 @@
 {% import '@SyliusUi/Macro/buttons.html.twig' as buttons %}
 
 <div class="ui top attached styled header">{{ 'sylius.ui.your_default_address'|trans }}</div>
-<div class="ui attached segment" id="sylius-default-address">
-    <div class="ui stackable two column grid">
+<div class="ui attached segment" id="sylius-default-address" {{ sylius_test_html_attribute('default-address') }}>
+    <div class="ui stackable two column grid" {{ sylius_test_html_attribute('address', address.firstName ~ ' ' ~ address.lastName) }}>
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Account/AddressBook/_defaultAddress.html.twig]: You missed this place :(

File: src/Sylius/Behat/Page/Shop/Account/AddressBook/IndexPage.php [hunk 11]
```diff
@@ -41,97 +34,61 @@ public function getAddressesCount()
...
     }
 
-    /**
-     * {@inheritdoc}
-     */
-    public function hasNoAddresses()
+    public function hasNoAddresses(): bool
     {
-        return $this->getDocument()->hasContent('You have no addresses defined');
+        return $this->hasElement('content');
```
Comment (GSadee) [inline src/Sylius/Behat/Page/Shop/Account/AddressBook/IndexPage.php]: You've missed checking if the proper message exists

File: src/Sylius/Behat/Page/Shop/Account/AddressBook/IndexPage.php [hunk 12]
```diff
@@ -41,97 +34,61 @@ public function getAddressesCount()
...
-    /**
-     * {@inheritdoc}
-     */
-    public function addressOfContains($fullName, $value)
+    public function addressOfContains(string $fullName, string $value): bool
     {
-        $address = $this->getAddressOf($fullName);
-
-        return $address->has('css', sprintf('address:contains("%s")', $value));
+         return null !== $this->getElement('address', ['%full_name%' => $fullName, '%value%' => $value]);
```
Comment (GSadee) [inline src/Sylius/Behat/Page/Shop/Account/AddressBook/IndexPage.php]:
```suggestion
         return $this->hasElement('address', ['%full_name%' => $fullName, '%value%' => $value]);
```

File: src/Sylius/Bundle/ShopBundle/Resources/views/Common/Form/_address.html.twig [hunk 13]
```diff
@@ -1,13 +1,15 @@
-{% if type is not defined %}
-     {% set type = null %}
+{% if type is defined %}
+    {% set type = type ~ '-' %}
```
Comment (GSadee) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Common/Form/_address.html.twig]: ?
IMO you should revert that

Comment (Tomanhez) [reply src/Sylius/Bundle/ShopBundle/Resources/views/Common/Form/_address.html.twig]: It was added by @AdamKasp for tests few days ago, It doesn't create BC-break, maybe we should sign it that it is using only for tests?

Comment (GSadee) [reply src/Sylius/Bundle/ShopBundle/Resources/views/Common/Form/_address.html.twig]: Probably I don't understand you, but you could revert changes in this file and everything should work well

**Comment by GSadee** [comment]
Thank you, Tomasz! :1st_place_medal:

File: src/Sylius/Bundle/ShopBundle/Resources/views/Common/Form/_address.html.twig (hunk 13 above)
Comment (lchrusciel) [reply src/Sylius/Bundle/ShopBundle/Resources/views/Common/Form/_address.html.twig]: Maybe the creation of a separate twig filter could make the biggest sense?

File: src/Sylius/Behat/Page/Shop/Account/AddressBook/UpdatePage.php [hunk 14]
```diff
@@ -90,37 +69,30 @@ public function selectCountry($name)
...
-            'street' => '#sylius_address_street',
+            'city' => '[data-test-city]',
+            'country' => '[data-test-country]',
+            'first_name' => '[data-test-first-name]',
+            'last_name' => '[data-test-last-name]',
+            'postcode' => '[data-test-postcode]',
+            'province_name' => '[data-test-province-name]',
+            'province_code' => '[data-test-province-code]',
+            'save_button' => '[data-test-save-changes]',
+            'selected_province' => '[data-test-province-code] option[selected="selected"]',
```
Comment (lchrusciel) [inline src/Sylius/Behat/Page/Shop/Account/AddressBook/UpdatePage.php]: I'm wondering if we shouldn't improve it and decouple from option element. Right now we are hardcoded to have a select box here.

/cc @kulczy

File: src/Sylius/Behat/Page/Shop/Account/AddressBook/CreatePage.php [hunk 15]
```diff
@@ -42,53 +36,39 @@ public function fillAddressData(AddressInterface $address)
...
-            'postcode' => '#sylius_address_postcode',
-            'street' => '#sylius_address_street',
+            'add_button' => '[data-test-add-address]',
+            'city' => '[data-test-city]',
+            'country' => '[data-test-country]',
+            'first_name' => '[data-test-first-name]',
+            'last_name' => '[data-test-last-name]',
+            'postcode' => '[data-test-postcode]',
+            'street' => '[data-test-street]',
+            'province_validation_message' => '[data-test-validation-error]:contains("province")',
```
Comment (lchrusciel) [inline src/Sylius/Behat/Page/Shop/Account/AddressBook/CreatePage.php]: Shouldn't we decouple more from the element itself? /cc @kulczy

**Pull request merged**